### PR TITLE
[FEATURE] PyProphetAdapter

### PR DESCRIPTION
**Pull request by singjc**
## Description

Added a new TOPP tool, PyProphetAdapter, for running PyProphet as part of the OpenSWATH workflow. This addition will hopefully make the whole OpenSWATH workflow easier, so that all the tools can be in a single place, especially for a user who maybe doesn't have as much experience with python (i.e. worrying about dependencies), and make it easier to integrate into workflows (KNIME, Nextflow, QuantMS).

I have created automated single-file binary executable builds with pyinstaller for pyprophet. However, the executables are kind of large, so git complains when trying to upload them to the THIRDPARTY repo, so I'm not sure what the best alternative would be to ship these with the THIRDPARTY stuff?

```
File MacOS/arm64/PyProphet/pyprophet is 233.37 MB
File MacOS/x86_64/PyProphet/pyprophet is 282.62 MB
File Windows/x86_64/PyProphet/pyprophet.exe is 220.27 MB
File Linux/x86_64/PyProphet/pyprophet is 312.53 MB
```

I have tried slimming down the installer and slimming down the dependencies within PyProphet to reduce the size (used to be 1.3Gb for Linux :grimacing: ...)

**PyProphetAdapter tool integration:**

* Registered `PyProphetAdapter` in the `tools_map` for tool handling, making it available as a recognized tool in OpenMS.
* Added `PyProphetAdapter` to the list of executables in `src/topp/executables.cmake`, ensuring it is built and distributed with OpenMS.

**Testing infrastructure:**

* Integrated detection of the `pyprophet` binary and added corresponding tests for `PyProphetAdapter` in `third_party_tests.cmake`, including checks for proper execution and error handling when the binary is missing. [[1]](diffhunk://#diff-4218a92bde2dbd1b648bf9149653c715ae656af87c7c256b38ed243d866ab687R91-R94) [[2]](diffhunk://#diff-4218a92bde2dbd1b648bf9149653c715ae656af87c7c256b38ed243d866ab687R293-R302)

**File path utility enhancements:**

* Added new static methods to `File.h`/`File.cpp`: `replaceExtension`, `appendSuffix`, and `stripExtension`, along with an internal helper to accurately identify file extensions, improving file name manipulation capabilities. [[1]](diffhunk://#diff-db9442289ec1f9a58a5dad7411cc13c2d928509b63c5ad43c8b30eed8357f3d1R323-R367) [[2]](diffhunk://#diff-e7d2d98e6bc31e659d7a9ed853b63d69a365b252c78fa83f0f51c8be4732cbecR968-R1007)

## Checklist
- [x] Make sure that you are listed in the AUTHORS file
- [x] Add relevant changes and new features to the CHANGELOG file
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] New and existing unit tests pass locally with my changes
- [x] Updated or added python bindings for changed or new classes (Tick if no updates were necessary.)

### How can I get additional information on failed tests during CI
<details>
  <summary>Click to expand</summary>
If your PR is failing you can check out

- The details of the action statuses at the end of the PR or the "Checks" tab.
- http://cdash.seqan.de/index.php?project=OpenMS and look for your PR. Use the "Show filters" capability on the top right to search for your PR number.
  If you click in the column that lists the failed tests you will get detailed error messages.

</details>

### Advanced commands (admins / reviewer only)
<details>
  <summary>Click to expand</summary>
  
- `/reformat` (experimental) applies the clang-format style changes as additional commit. Note: your branch must have a different name (e.g., yourrepo:feature/XYZ) than the receiving branch (e.g., OpenMS:develop). Otherwise, reformat fails to push.
- setting the label "NoJenkins" will skip tests for this PR on jenkins (saves resources e.g., on edits that do not affect tests)
- commenting with `rebuild jenkins` will retrigger Jenkins-based CI builds
  
</details>

---
:warning: Note: Once you opened a PR try to minimize the number of *pushes* to it as every push will trigger CI (automated builds and test) and is rather heavy on our infrastructure (e.g., if several pushes per day are performed).

## Summary by CodeRabbit

* **New Features**
  * PyProphetAdapter: merge OSW inputs, configurable multi-level scoring with several classifiers, peptide/protein/gene inference, peptidoform inference (IPF), and exports (TSV, proteomics matrices, reports, plots).
  * File-path utilities: replace/strip extensions and insert filename suffixes.

* **Tests**
  * Added tests for PyProphetAdapter output validation, a missing-executable scenario, and output diff verification.

### CODE REVIEW

**Comment by singjc** [comment]
@timosachsenberg , can you please take a look and let me know if I missed anything? I added tests to the thirdy party tests, but not sure if I missed anything?

**Comment by jpfeuffer** [comment]
Hmmm personally I would have done that different. 

With uv and pixi, no one ever has to install python anymore. You can even encode your requirements at the top of your python file as a comment. UV or pixi are single line installs on both windows and Linux.
The installers could even run a post install script to download uv and construct a venv nested somewhere in the I stallation tree.

Even better, why not rewrite this thing in a compiled language that doesnt need 1GB of interpreter etc.
It shouldn't be too hard in the AI days to translate this into ideally C++ to include it here, but could be Rust for all I care.

To be fair, your solution is one of the safest and definitely better than letting the user install everything but IMHO it is definitely too large to ship by default.

**Comment by singjc** [comment]
> Hmmm personally I would have done that different.
> 
> With uv and pixi, no one ever has to install python anymore. You can even encode your requirements at the top of your python file as a comment. UV or pixi are single line installs on both windows and Linux. The installers could even run a post install script to download uv and construct a venv nested somewhere in the I stallation tree.

Thanks for the feedback! 

I didn't think about packaging with uv! If we went down the route of uv/pixi, would we just have a general installer script that would download uv, create the venv and installing a python package target (and associated dependencies) from the THIRDPARTY repo?

> Even better, why not rewrite this thing in a compiled language that doesnt need 1GB of interpreter etc. It shouldn't be too hard in the AI days to translate this into ideally C++ to include it here, but could be Rust for all I care.

I have thought about re-writing / transferring some of PyProphet to a rust library that I have that already does PSM rescoring, but I left it as a thought haha. Maybe I will revise to see how feasible it would be.

> To be fair, your solution is one of the safest and definitely better than letting the user install everything but IMHO it is definitely too large to ship by default.

I agree, the current binary is a little too large to ship. Which is mainly due to some larger dependencies that it relies on.

* pyarrow (133M) + arrow 
* polars (106M) (we were initially thinking of replacing pandas with polars)
* pyopenms (80M) + pyopenms.libs (94M)
* duckdb (61M)
* pandas (46M) + numpy/scipy/sklearn (~150M combined)

I am trying to remove/reduce some of the dependencies, could probably make parquet/pyarrow as an optional dependency since the adapter will currently only accept sqlite-OSW files.

Something that I do wish was possible with pyopenms, which maybe you you know if it's possible or not, is if it was split into optional dependency parts (i.e. pyopenms[core-chemistry stuff], pyopenms[low-level algo stuff], pyopenms[file i/o stuff] etc). Because we only require pyopenms as a dependency in PyProphet to use the `AASequence` class for converting between UniMod and Codenames.

**Review by jcharkow — COMMENTED** [review]
In terms of the code everything looks good.  I like the idea of having everything in one place however I do not know how much it would be used in practice. It seems a bit clunky that we need to create a python installer and link with OpenMS. 


If pyprophet is truly stable, I think the most sense would be to implement the algorithm within OpenMS (even just for SVM), but that does make it harder for python developers. Furthermore, pyprophet already works quite well in python because of all the machine learning. Also, you just did a lot of good refactoring for pyprophet, so it would be a waste if we just move everything back to c++.


Also if we wanted things to truly be together it might make more sense to just have a single executable which does both openswathworkflow and pyprophet in one step? 


Sorry for the ramble, in summary I'm not sure of this is the best solution at the current stage of development but that's just my opinion.

**Comment by singjc** [comment]
> In terms of the code everything looks good. I like the idea of having everything in one place however I do not know how much it would be used in practice. It seems a bit clunky that we need to create a python installer and link with OpenMS.

Thanks for the feedback!

This is kind of why I went down the route of creating a portable single binary executable for pyprophet, so it can be shipped with OpenMS via the thirdparty tools like Percolator, Sage, MSFragger etc. The issue is it's a little too large, due to some of it's dependencies. A uv installer script approach might be one way around this, because it could deal with all the python and venv management behind the scenes for the user. This could also make it more tractable for including other python tools, like Mokapot, which I think was also in discussion for including in the thirdparty repo. 

> If pyprophet is truly stable, I think the most sense would be to implement the algorithm within OpenMS (even just for SVM), but that does make it harder for python developers. Furthermore, pyprophet already works quite well in python because of all the machine learning. Also, you just did a lot of good refactoring for pyprophet, so it would be a waste if we just move everything back to c++.

I do agree, I think pyprophet works well in python because we can more easily test and use different classifiers for scoring from sklearn. I would be disheartened if all the refactoring went to waste :smiling_face_with_tear:. I was thinking of re-writing some of the heavy stuff in rust/c++ and cythonizing it to reduce some of the dependencies. There are a few functions from scipy stats, that I think we could reproduce/reimplement to avoiding having it as a dependency, since we don't heavily rely on it. We mostly depend on pandas, numpy, sklearn and duckdb (with the recent changes).

Reducing some of the larger dependencies will help reduce the pyinstaller produced binary.

> Also if we wanted things to truly be together it might make more sense to just have a single executable which does both openswathworkflow and pyprophet in one step?

This was also part of the motivation by adding the adapter, because we could then call the adapter at the end of the openswathworkflow. However, the caveat then would be that it can only operate on single runs, because currently openswathworkflow only works on either a single mzML file or split mzML file (same experiment, split by swaths). Otherwise, we would have to modify OSW to accept multiple experiment mzML files, and perform the feature finding and feature scoring sequentially per file, and then call the pyprophet adapter. I don't think this would be too ideal though, time wise, unless we can implement some sort of automated distributed computing if detected.

**Comment by timosachsenberg** [comment]
To address the binary size issue, I think the most reasonable approach would be to translate the code to C++ or Rust. This would also simplify deployment. In fact, the easiest option might be to integrate it directly as an external library within the OpenMS source tree, which would also allow us to remove the circular dependency back to (py)OpenMS.

That would be my preferred solution, although keeping it synchronized could be challenging if significant development is still happening in the Python script. From what it sounds like, Justin expects further changes there, so I think the current setup remains a sensible choice for now.

@jpfeuffer — how would the UV integration interact with our THIRDPARTY setup? Would we store the related information there, or rather within the OpenMS/OpenMS repository? How would this play with e.g. conda THIRDPARTY?

**Comment by jpfeuffer** [comment]
First of all, I think only KNIME is a real problem also. For nextflow (incl. quantms) everything is containerized or based on Conda environments anyway. We should add a way to alternatively point this adapter here to a Python.exe of such an environment instead of a bundled pyinstaller exe.

For users of KNIME or other non container workflows:

I think we would need to try how a UV install script works out.
- If users have enough permissions (I think so because you can download uv to anywhere and create venvs everywhere, incl. in user space)
- If users are okay with requiring downloads after installation (I guess everyone but high security environments will be fine)
- If we do it post install or in the tool

And compare it to the trouble of potentially offering an optional installer that just installs python components of OpenMS (for now, Py prophet), if you can't get it "small enough". Now that I am thinking, KNIME would be a real trouble indeed because it would need to be another plug-in if that part is optional.
Alternatively, one could also test how difficult it  actually is to create a python environment in knime and use it in a downstream node. KNIME made this very easy in the last years.

In general, one thing to consider is also to write the whole TOPP tool in python. There is a python implementation of our CTD/ini-based CLInterface. We just need to make those python based tools discoverable and have them in the documentation. There were attempts before but only by students with not enough time to complete. An experienced developer like you might be able to do that quickly. @timosachsenberg probably knows more.

**Comment by jpfeuffer** [comment]
> @jpfeuffer — how would the UV integration interact with our THIRDPARTY setup? Would we store the related information there, or rather within the OpenMS/OpenMS repository? How would this play with e.g. conda THIRDPARTY?

I think we would make the adapter point to a python.exe of a venv (Conda or not does not matter). For our conda package we could make sure it points to the python exe of the env it was installed into by default (not sure how but certainly possible).

For installers, we could create a venv in the openms folder with a post script or at runtime (if permissions allow?).

The problem would be if we start to go full cycle and have python tools that require pyopenms 🤯 (but then it wouldn't really be a third-party tool anymore)

**Comment by timosachsenberg** [comment]
> In general, one thing to consider is also to write the whole TOPP tool in python. There is a python implementation of our CTD/ini-based CLInterface. We just need to make those python based tools discoverable and have them in the documentation. There were attempts before but only by students with not enough time to complete. An experienced developer like you might be able to do that quickly. @timosachsenberg probably knows more.

There are a few examples in the pyTOPP folder:
https://github.com/OpenMS/OpenMS/blob/develop/src/pyOpenMS/pyTOPP/pyTOPPExample.py
what was not dealt/discussed:
- the extra dependencies these tools might introduce and how to specify them (e.g. per tool requirements, at pyTOPP level, is there a better way ?).
- How to integrate them into our build system such that (1) documentation is build, (2) they can be tested, (3) they are registered as a TOPPTool, (4) they can be deployed

**Comment by jpfeuffer** [comment]
Thanks! If we continue this I think we should start with tool level deps. With uv you can spawn a venv in a second.
It is safer to not have version clashes.

**Comment by timosachsenberg** [comment]
@jpfeuffer do you think it is worthwhile to give this a try for the PyProphetAdapter?
How would a call to the tool look like?
ideally, we could just call the tool as a normal executable and it would install dependencies on first run etc.

**Comment by jpfeuffer** [comment]
PyTopp tool:

```python
#!/usr/bin/env -S uv run --script
# /// script
# dependencies = ["ctdopts", "pyprophet"]
# ///

import ctdopts as ctd
import pyprophet as pp

# define ctd interface ...
# run pyprophet
# ???
# profit
```

This just requires uv. And can be run (after making it executable) with ./PyProphet.py

**Comment by timosachsenberg** [comment]
@singjc would this be something you'd be interested in trying out? I think the adapter could more or less be translated automatically to pyTOPP.
Sorry if this adds some extra work, but moving this forward would really strengthen our ability to develop tools directly in Python.

**Comment by timosachsenberg** [comment]
@jpfeuffer can you confirm this is a good way forward or will we hit troubles when deploying this in conda?

**Comment by jpfeuffer** [comment]
I think even Conda should work because uv will create new environments for every pytopp tool then.

**Comment by singjc** [comment]
Thanks for the great discussion and suggestions! I can give creating a pyTopp tool a go.

**Comment by singjc** [comment]
Will close in favour of the pytopp direction.
